### PR TITLE
`Development`: Fix team students eager loading for several occasions

### PR DESCRIPTION
**Pull request by julian-christl**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [X] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [X] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [X] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [X] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [X] I documented the Java code using JavaDoc style.
#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server configured with **Gitlab** and **Jenkins**.
- Removed Atlassian Stack as it's deprecated

### Motivation and Context

After issues with #7829 and some hot fixes, I went through the whole code base to find any occurrence where loading team students might be necessary. It took some time, but I was able to go through most branches. There were a few that I had to skip because they were just too complex, but I added an initialisation on these instances, thus they are still covered.

### Description

Eagerly loaded team students wherever they are called. In some areas I maintained the relations after saving by overriding them afterwards. Some queries I had to extract as `participation.team.students` won't work on the entity graph because team is on the student participation, not the abstract one. In some sections I load the students if not initialised. Usually because it would be too much effort to go through all call branches and make sure that every one of them loads the students for team exercises. Finally I also modified some sections to reduce database calls and improved some queries.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 2 Students in a team
- 1 Exercise of each type as team exercise

1. Ensure that the team administration (create, update, etc) is fully working. You can check this in the test preparation already.
2. Do all steps of the participation (submit, assess, complain, answer complain) and make sure that for every step you check all team views (i.e. the view where team members are listed).
3. Make sure that the scores are displayed correctly at all times for students and instructor alike
4. Make sure that when updating the due date of programming exercises, the repositories get locked/unlocked correctly
5. Make sure that the websocket events work especially for the students and hence the values get updated. If you're not sure what to expect, a good suggestion would be to monitor sentry during your tests and make sure that there are no errors from your server related to this PR (especially initialization errors) https://sentry.ase.in.tum.de/organizations/artemis/issues/?environment=test&project=2&statsPeriod=1h

#### Exam Mode Testing

Not required, exams don't support team exercises.

#### Performance Review
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

No functional changes really affecting the coverage. If you really need the coverage, let me know and I'll add it.

## Summary by CodeRabbit

- **Refactor**
	- Improved efficiency in fetching related entities for exercises.
	- Enhanced submission query methods for better data handling and added a new method for fetching specific submission associations.

### CODE REVIEW

**Comment by krusche** [comment]
I am not sure if the PR follows the right approach. It seems we load team and students way too often now. I guess it would be better to only load the team with students, when we actually have a team exercise

**Comment by krusche** [comment]
I made sure that the hot fix in the release branch `release/6.7.7` was merged into `develop` to avoid a regression in 6.8.0

**Comment by julian-christl** [comment]
> I am not sure if the PR follows the right approach. It seems we load team and students way too often now. I guess it would be better to only load the team with students, when we actually have a team exercise

So do I get you right that you would prefer having two different repository methods for most of these methods and call the corresponding one depending on whether `exercise.isTeamMode()` resolves to false or true?
I don't see the benefit. If it's not a team exercise, it won't get loaded anyway.

Or did I misinterpret your intention here?

File: src/main/java/de/tum/in/www1/artemis/service/scheduled/cache/quiz/QuizScheduleService.java [hunk 1]
```diff
@@ -506,9 +512,10 @@ public void processCachedQuizSubmissions() {
                             log.error("Participation is missing student (or student is missing username): {}", participation);
                         }
                         else {
-                            if(ltiNewResultService.isPresent()) {
-                                ltiNewResultService.get().onNewResult(participation);
+                            if (participation.getParticipant() instanceof Team team && !Hibernate.isInitialized(team.getStudents())) {
+                                participation.setParticipant(teamRepository.findWithStudentsByIdElseThrow(team.getId()));
                             }
+                            ltiNewResultService.ifPresent(newResultService->newResultService.onNewResult(participation));
```
[the remark this answers is not in the document]
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/service/scheduled/cache/quiz/QuizScheduleService.java:518]: It's just one additional call in that case and it's a rare endpoint used so there is no performance concern on my side

File: src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java [hunk 2]
```diff
@@ -502,16 +505,44 @@ public ResponseEntity<List<StudentParticipation>> updateParticipationDueDates(@P
...
+            return;
+        }
+        participations.stream().filter(exerciseDateService::isBeforeDueDate).forEach(participation -> {
+            var studentParticipation = (ProgrammingExerciseStudentParticipation) participation;
+            if (studentParticipation.getParticipant() instanceof Team team && !Hibernate.isInitialized(team.getStudents())) {
+                studentParticipation.setParticipant(teamRepository.findWithStudentsByIdElseThrow(team.getId()));
+            }
+            programmingExerciseParticipationService.unlockStudentRepositoryAndParticipation((ProgrammingExerciseStudentParticipation) participation);
+        });
+    }
```
[the remark this answers is not in the document]
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java]: Fixed in 3e35a0941f104c356dca6c423054b9dad78e67ed

File: src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java [hunk 3]
```diff
@@ -502,16 +505,44 @@
...
+            return;
+        }
+        participations.stream().filter(exerciseDateService::isAfterDueDate).forEach(participation -> {
+            var studentParticipation = (ProgrammingExerciseStudentParticipation) participation;
+            if (studentParticipation.getParticipant() instanceof Team team && !Hibernate.isInitialized(team.getStudents())) {
+                studentParticipation.setParticipant(teamRepository.findWithStudentsByIdElseThrow(team.getId()));
+            }
+            programmingExerciseParticipationService.lockStudentRepositoryAndParticipation(exercise, (ProgrammingExerciseStudentParticipation) participation);
+        });
+    }
```
[the remark this answers is not in the document]
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java]: Fixed in 3e35a0941f104c356dca6c423054b9dad78e67ed

**Review by undernagruzez — APPROVED** [review]
code lgtm
tested on ts3
exercise & team creation works as expected
could change the due date to the past and repos got locked
changed the due date back to the future, but the repos remained locks, not sure if it is the expected behavior
did the assessment manually, the students were not notified about the results, also not sure it they should be
also, found these exceptions on sentry: https://sentry.ase.in.tum.de/organizations/artemis/issues/37553/?environment=test&project=2&query=is%3Aunresolved&referrer=issue-stream&statsPeriod=1h
https://sentry.ase.in.tum.de/organizations/artemis/issues/40139/?environment=test&project=2&query=is%3Aunresolved&referrer=issue-stream&statsPeriod=1h
https://sentry.ase.in.tum.de/organizations/artemis/issues/26185/?environment=test&project=2&query=is%3Aunresolved&referrer=issue-stream&statsPeriod=1h

**Comment by julian-christl** [comment]
> code lgtm tested on ts3 exercise & team creation works as expected could change the due date to the past and repos got locked changed the due date back to the future, but the repos remained locked, not sure if it is the expected behavior did the assessment manually, the students were not notified about the results, also not sure it they should be also, found these exceptions on sentry: https://sentry.ase.in.tum.de/organizations/artemis/issues/37553/?environment=test&project=2&query=is%3Aunresolved&referrer=issue-stream&statsPeriod=1h https://sentry.ase.in.tum.de/organizations/artemis/issues/40139/?environment=test&project=2&query=is%3Aunresolved&referrer=issue-stream&statsPeriod=1h https://sentry.ase.in.tum.de/organizations/artemis/issues/26185/?environment=test&project=2&query=is%3Aunresolved&referrer=issue-stream&statsPeriod=1h

Thank you for testing
1. Unlocking is not supported, I updated the testing instructions accordingly
2. They should not be notified in terms of a notification, but the result should be updated and reflect the new percentage.
3. I fixed the nullpointer exceptions. I'd appreciate if you could recheck.
4. The last exception is unrelated.

**Review by undernagruzez — APPROVED** [review]
tested again on ts3
didn't encounter any errors

**Review by coolchock — DISMISSED** [review]
tested everything on ts3, works as expected. It is possible to submit, assess, complain and respond to a complaint. Websockets work too, students get notified about a new assessment. Code LGTM.

File: src/main/java/de/tum/in/www1/artemis/service/notifications/SingleUserNotificationService.java [hunk 4]
```diff
@@ -117,6 +118,7 @@ private SingleUserNotification createSingleUserNotification(Object notificationS
      *
      * @param exercise which assessmentDueDate is the trigger for the notification process
      */
+    // TODO: Should by a general method and not be in the single user service
```
Comment (rstief) [inline src/main/java/de/tum/in/www1/artemis/service/notifications/SingleUserNotificationService.java:121]: do this in a follow-up or create an issue for this, so we don't forget it? :)

**Review by rstief — DISMISSED** [review]
Code looks good 👍

**Review by basak-akan — DISMISSED** [review]
Tested on TS3, followed testing steps works as expected

**Review by mateusmm01 — COMMENTED** [review]
unrelated

**Review by mateusmm01 — APPROVED** [review]
Tested on ts5 and works as expected 🚀

**Review by rstief — APPROVED** [review]
reapprove

**Review by coolchock — APPROVED** [review]

**Review by basak-akan — APPROVED** [review]
Re approve

**Review by undernagruzez — APPROVED** [review]
Re approve

**Pull request merged**